File: external/boost_1.55.0/libs/python/test/test_builtin_converters.py
# Copyright David Abrahams 2004. Distributed under the Boost
# Software License, Version 1.0. (See accompanying
# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
r"""
>>> from builtin_converters_ext import *

# Provide values for integer converter tests
>>> def _signed_values(s):
...     base = 2 ** (8 * s - 1)
...     return [[-base, -1, 1, base - 1], [-base - 1, base]]
>>> def _unsigned_values(s):
...     base = 2 ** (8 * s)
...     return [[1, base - 1], [-1L, -1, base]]

# Wrappers to simplify tests
>>> def should_pass(method, values):
...     result = map(method, values[0])
...     if result != values[0]:
...         print "Got %s but expected %s" % (result, values[0])
>>> def test_overflow(method, values):
...     for v in values[1]:
...         try: method(v)
...         except OverflowError: pass
...         else: print "OverflowError expected"

# Synthesize idendity functions in case long long not supported
>>> if not 'rewrap_value_long_long' in dir():
...     def rewrap_value_long_long(x): return long(x)
...     def rewrap_value_unsigned_long_long(x): return long(x)
...     def rewrap_const_reference_long_long(x): return long(x)
...     def rewrap_const_reference_unsigned_long_long(x): return long(x)
>>> if not 'long_long_size' in dir():
...     def long_long_size(): return long_size()

>>> try: bool_exists = bool
... except: pass
... else:
...     rewrap_value_bool(True)
...     rewrap_value_bool(False)
True
False

>>> rewrap_value_bool(None)
0
>>> rewrap_value_bool(0)
0
>>> rewrap_value_bool(33)
1
>>> rewrap_value_char('x')
'x'

  Note that there's currently silent truncation of strings passed to
  char arguments.

>>> rewrap_value_char('xy')
'x'
>>> rewrap_value_signed_char(42)
42
>>> rewrap_value_unsigned_char(42)
42
>>> rewrap_value_int(42)
42
>>> rewrap_value_unsigned_int(42)
42
>>> rewrap_value_short(42)
42
>>> rewrap_value_unsigned_short(42)
42
>>> rewrap_value_long(42)
42
>>> rewrap_value_unsigned_long(42)
42

    test unsigned long values which don't fit in a signed long.
    strip any 'L' characters in case the platform has > 32 bit longs
        
>>> hex(rewrap_value_unsigned_long(0x80000001L)).replace('L','')
'0x80000001'

>>> rewrap_value_long_long(42) == 42
True
>>> rewrap_value_unsigned_long_long(42) == 42
True

   show that we have range checking. 

>>> should_pass(rewrap_value_signed_char, _signed_values(char_size()))
>>> should_pass(rewrap_value_short, _signed_values(short_size()))
>>> should_pass(rewrap_value_int, _signed_values(int_size()))
>>> should_pass(rewrap_value_long, _signed_values(long_size()))
>>> should_pass(rewrap_value_long_long, _signed_values(long_long_size()))

>>> should_pass(rewrap_value_unsigned_char, _unsigned_values(char_size()))
>>> should_pass(rewrap_value_unsigned_short, _unsigned_values(short_size()))
>>> should_pass(rewrap_value_unsigned_int, _unsigned_values(int_size()))
>>> should_pass(rewrap_value_unsigned_long, _unsigned_values(long_size()))
>>> should_pass(rewrap_value_unsigned_long_long,
...     _unsigned_values(long_long_size()))

>>> test_overflow(rewrap_value_signed_char, _signed_values(char_size()))
>>> test_overflow(rewrap_value_short, _signed_values(short_size()))
>>> test_overflow(rewrap_value_int, _signed_values(int_size()))
>>> test_overflow(rewrap_value_long, _signed_values(long_size()))
>>> test_overflow(rewrap_value_long_long, _signed_values(long_long_size()))

>>> test_overflow(rewrap_value_unsigned_char, _unsigned_values(char_size()))
>>> test_overflow(rewrap_value_unsigned_short, _unsigned_values(short_size()))
>>> test_overflow(rewrap_value_unsigned_int, _unsigned_values(int_size()))
>>> test_overflow(rewrap_value_unsigned_long, _unsigned_values(long_size()))

# Exceptionally for PyLong_AsUnsignedLongLong(), a negative value raises
# TypeError on Python versions prior to 2.7
>>> for v in _unsigned_values(long_long_size())[1]:
...     try: rewrap_value_unsigned_long_long(v)
...     except (OverflowError, TypeError): pass
...     else: print "OverflowError or TypeError expected"

>>> assert abs(rewrap_value_float(4.2) - 4.2) < .000001
>>> rewrap_value_double(4.2) - 4.2
0.0
>>> rewrap_value_long_double(4.2) - 4.2
0.0

>>> assert abs(rewrap_value_complex_float(4+.2j) - (4+.2j)) < .000001
>>> assert abs(rewrap_value_complex_double(4+.2j) - (4+.2j)) < .000001
>>> assert abs(rewrap_value_complex_long_double(4+.2j) - (4+.2j)) < .000001

>>> rewrap_value_cstring('hello, world')
'hello, world'
>>> rewrap_value_string('yo, wassup?')
'yo, wassup?'

>>> print rewrap_value_wstring(u'yo, wassup?')
yo, wassup?

   test that overloading on unicode works:

>>> print rewrap_value_string(u'yo, wassup?')
yo, wassup?

   wrap strings with embedded nulls:
   
>>> rewrap_value_string('yo,\0wassup?')
'yo,\x00wassup?'

>>> rewrap_value_handle(1)
1
>>> x = 'hi'
>>> assert rewrap_value_handle(x) is x
>>> assert rewrap_value_object(x) is x

  Note that we can currently get a mutable pointer into an immutable
  Python string:
  
>>> rewrap_value_mutable_cstring('hello, world')
'hello, world'

>>> rewrap_const_reference_bool(None)
0
>>> rewrap_const_reference_bool(0)
0

>>> try: rewrap_const_reference_bool('yes')
... except TypeError: pass
... else: print 'expected a TypeError exception'

>>> rewrap_const_reference_char('x')
'x'

  Note that there's currently silent truncation of strings passed to
  char arguments.

>>> rewrap_const_reference_char('xy')
'x'
>>> rewrap_const_reference_signed_char(42)
42
>>> rewrap_const_reference_unsigned_char(42)
42
>>> rewrap_const_reference_int(42)
42
>>> rewrap_const_reference_unsigned_int(42)
42
>>> rewrap_const_reference_short(42)
42
>>> rewrap_const_reference_unsigned_short(42)
42
>>> rewrap_const_reference_long(42)
42
>>> rewrap_const_reference_unsigned_long(42)
42
>>> rewrap_const_reference_long_long(42) == 42
True
>>> rewrap_const_reference_unsigned_long_long(42) == 42
True


>>> assert abs(rewrap_const_reference_float(4.2) - 4.2) < .000001
>>> rewrap_const_reference_double(4.2) - 4.2
0.0
>>> rewrap_const_reference_long_double(4.2) - 4.2
0.0

>>> assert abs(rewrap_const_reference_complex_float(4+.2j) - (4+.2j)) < .000001
>>> assert abs(rewrap_const_reference_complex_double(4+.2j) - (4+.2j)) < .000001
>>> assert abs(rewrap_const_reference_complex_long_double(4+.2j) - (4+.2j)) < .000001

>>> rewrap_const_reference_cstring('hello, world')
'hello, world'
>>> rewrap_const_reference_string('yo, wassup?')
'yo, wassup?'

>>> rewrap_const_reference_handle(1)
1
>>> x = 'hi'
>>> assert rewrap_const_reference_handle(x) is x
>>> assert rewrap_const_reference_object(x) is x
>>> assert rewrap_reference_object(x) is x


Check that None <==> NULL

>>> rewrap_const_reference_cstring(None)

But None cannot be converted to a string object:

>>> try: rewrap_const_reference_string(None)
... except TypeError: pass
... else: print 'expected a TypeError exception'

Now check implicit conversions between floating/integer types

>>> rewrap_const_reference_float(42)
42.0

>>> rewrap_const_reference_float(42L)
42.0

>>> try: rewrap_const_reference_int(42.0)
... except TypeError: pass
... else: print 'expected a TypeError exception'

>>> rewrap_value_float(42)
42.0

>>> try: rewrap_value_int(42.0)
... except TypeError: pass
... else: print 'expected a TypeError exception'

Check that classic classes also work

>>> class FortyTwo:
...     def __int__(self):
...         return 42
...     def __float__(self):
...         return 42.0
...     def __complex__(self):
...         return complex(4+.2j)
...     def __str__(self):
...         return '42'

>>> try: rewrap_const_reference_float(FortyTwo())
... except TypeError: pass
... else: print 'expected a TypeError exception'

>>> try: rewrap_value_int(FortyTwo())
... except TypeError: pass
... else: print 'expected a TypeError exception'

>>> try: rewrap_const_reference_string(FortyTwo())
... except TypeError: pass
... else: print 'expected a TypeError exception'

>>> try: rewrap_value_complex_double(FortyTwo())
... except TypeError: pass
... else: print 'expected a TypeError exception'

# show that arbitrary handle<T> instantiations can be returned
>>> assert get_type(1) is type(1)

>>> assert return_null_handle() is None
"""

def run(args = None):
    import sys
    import doctest
    import builtin_converters_ext
    
    if 'rewrap_value_long_long' in dir(builtin_converters_ext):
        print 'LONG_LONG supported, testing...'
    else:
        print 'LONG_LONG not supported, skipping those tests...'
        
    if args is not None:
        sys.argv = args
    return doctest.testmod(sys.modules.get(__name__))
    
if __name__ == '__main__':
    print "running..."
    import sys
    status = run()[0]
    if (status == 0): print "Done."
    sys.exit(status)

File: external/boost_1.55.0/libs/python/test/test_builtin_converters.cpp
// Copyright David Abrahams 2002.
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
#include <boost/python/module.hpp>
#include <boost/python/def.hpp>
#include <complex>
#include <boost/python/handle.hpp>
#include <boost/python/cast.hpp>
#include <boost/python/object.hpp>
#include <boost/python/detail/wrap_python.hpp>

template <class T>
struct by_value
{
    static T rewrap(T x)
    {
        return x;
    }
    static int size(void)
    {
        return sizeof(T);
    }
};

template <class T>
struct by_const_reference
{
    static T rewrap(T const& x)
    {
        return x;
    }
};

template <class T>
struct by_reference
{
    static T rewrap(T& x)
    {
        return x;
    }
};

using boost::python::def;
using boost::python::handle;
using boost::python::object;
using boost::python::borrowed;

// Used to test that arbitrary handle<>s can be returned
handle<PyTypeObject> get_type(handle<> x)
{
    return handle<PyTypeObject>(borrowed(x->ob_type));
}

handle<> return_null_handle()
{
    return handle<>();
}

char const* rewrap_value_mutable_cstring(char* x) { return x; }

object identity_(object x) { return x; }

BOOST_PYTHON_MODULE(builtin_converters_ext)
{    
    def("get_type", get_type);
    def("return_null_handle", return_null_handle);

// These methods are used solely for getting some C++ type sizes
    def("bool_size", by_value<bool>::size);
    def("char_size", by_value<char>::size);
    def("int_size", by_value<int>::size);
    def("short_size", by_value<short>::size);
    def("long_size", by_value<long>::size);
#ifdef HAVE_LONG_LONG
    def("long_long_size", by_value<BOOST_PYTHON_LONG_LONG>::size);
#endif

    def("rewrap_value_bool", by_value<bool>::rewrap);
    def("rewrap_value_char", by_value<char>::rewrap);
    def("rewrap_value_signed_char", by_value<signed char>::rewrap);
    def("rewrap_value_unsigned_char", by_value<unsigned char>::rewrap);
    def("rewrap_value_int", by_value<int>::rewrap);
    def("rewrap_value_unsigned_int", by_value<unsigned int>::rewrap);
    def("rewrap_value_short", by_value<short>::rewrap);
    def("rewrap_value_unsigned_short", by_value<unsigned short>::rewrap);
    def("rewrap_value_long", by_value<long>::rewrap);
    def("rewrap_value_unsigned_long", by_value<unsigned long>::rewrap);
// using Python's macro instead of Boost's - we don't seem to get the
// config right all the time.
#ifdef HAVE_LONG_LONG
    def("rewrap_value_long_long", by_value<BOOST_PYTHON_LONG_LONG>::rewrap);
    def("rewrap_value_unsigned_long_long", by_value<unsigned BOOST_PYTHON_LONG_LONG>::rewrap);
# endif 
    def("rewrap_value_float", by_value<float>::rewrap);
    def("rewrap_value_double", by_value<double>::rewrap);
    def("rewrap_value_long_double", by_value<long double>::rewrap);
    def("rewrap_value_complex_float", by_value<std::complex<float> >::rewrap);
    def("rewrap_value_complex_double", by_value<std::complex<double> >::rewrap);
    def("rewrap_value_complex_long_double", by_value<std::complex<long double> >::rewrap);
    def("rewrap_value_wstring",
# if defined(BOOST_NO_STD_WSTRING) || !defined(Py_USING_UNICODE)
        identity_
# else 
        by_value<std::wstring>::rewrap
# endif 
    );
    def("rewrap_value_string",
# if defined(BOOST_NO_STD_WSTRING) || !defined(Py_USING_UNICODE)
        identity_
# else 
        by_value<std::wstring>::rewrap
# endif 
    );
    def("rewrap_value_string", by_value<std::string>::rewrap);
    def("rewrap_value_cstring", by_value<char const*>::rewrap);
    def("rewrap_value_handle", by_value<handle<> >::rewrap);
    def("rewrap_value_object", by_value<object>::rewrap);

        // Expose this to illustrate our failings ;-). See test_builtin_converters.py
    def("rewrap_value_mutable_cstring", rewrap_value_mutable_cstring);


    def("rewrap_const_reference_bool", by_const_reference<bool>::rewrap);
    def("rewrap_const_reference_char", by_const_reference<char>::rewrap);
    def("rewrap_const_reference_signed_char", by_const_reference<signed char>::rewrap);
    def("rewrap_const_reference_unsigned_char", by_const_reference<unsigned char>::rewrap);
    def("rewrap_const_reference_int", by_const_reference<int>::rewrap);
    def("rewrap_const_reference_unsigned_int", by_const_reference<unsigned int>::rewrap);
    def("rewrap_const_reference_short", by_const_reference<short>::rewrap);
    def("rewrap_const_reference_unsigned_short", by_const_reference<unsigned short>::rewrap);
    def("rewrap_const_reference_long", by_const_reference<long>::rewrap);
    def("rewrap_const_reference_unsigned_long", by_const_reference<unsigned long>::rewrap);
// using Python's macro instead of Boost's - we don't seem to get the
// config right all the time.
# ifdef HAVE_LONG_LONG
    def("rewrap_const_reference_long_long", by_const_reference<BOOST_PYTHON_LONG_LONG>::rewrap);
    def("rewrap_const_reference_unsigned_long_long", by_const_reference<unsigned BOOST_PYTHON_LONG_LONG>::rewrap);
# endif
    def("rewrap_const_reference_float", by_const_reference<float>::rewrap);
    def("rewrap_const_reference_double", by_const_reference<double>::rewrap);
    def("rewrap_const_reference_long_double", by_const_reference<long double>::rewrap);
    def("rewrap_const_reference_complex_float", by_const_reference<std::complex<float> >::rewrap);
    def("rewrap_const_reference_complex_double", by_const_reference<std::complex<double> >::rewrap);
    def("rewrap_const_reference_complex_long_double", by_const_reference<std::complex<long double> >::rewrap);
    def("rewrap_const_reference_string", by_const_reference<std::string>::rewrap);
    def("rewrap_const_reference_cstring", by_const_reference<char const*>::rewrap);
    def("rewrap_const_reference_handle", by_const_reference<handle<> >::rewrap);
    def("rewrap_const_reference_object", by_const_reference<object>::rewrap);
    def("rewrap_reference_object", by_reference<object>::rewrap);
}

